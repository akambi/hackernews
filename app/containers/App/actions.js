/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  LOAD_STORIES,
  LOAD_STORIES_SUCCESS,
  LOAD_STORIES_ERROR,
} from './constants';

/**
 * Load the stories, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_STORIES
 */
export function loadStories() {
  return {
    type: LOAD_STORIES,
  };
}

/**
 * Dispatched when the stories are loaded by the request saga
 *
 * @param  {array} stories The story data
 * @param  {string} searchQuery The current searchQuery
 *
 * @return {object}      An action object with a type of LOAD_STORIES_SUCCESS passing the stories
 */
export function storiesLoaded(stories, searchQuery) {
  return {
    type: LOAD_STORIES_SUCCESS,
    stories,
    searchQuery,
  };
}

/**
 * Dispatched when loading the stories fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_STORIES_ERROR passing the error
 */
export function repoLoadingError(error) {
  return {
    type: LOAD_STORIES_ERROR,
    error,
  };
}
