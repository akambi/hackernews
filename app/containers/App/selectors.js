/**
 * The global state selectors
 */

import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectGlobal = state => state.global || initialState;

const selectRouter = state => state.router;

const makeSelectCurrentUser = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.currentSearchQuery,
  );

const makeSelectLoading = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.loading,
  );

const makeSelectError = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.error,
  );

const makeSelectStories = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.userData.stories,
  );

const makeSelectLocation = () =>
  createSelector(
    selectRouter,
    routerState => routerState.location,
  );

export {
  selectGlobal,
  makeSelectCurrentUser,
  makeSelectLoading,
  makeSelectError,
  makeSelectStories,
  makeSelectLocation,
};
