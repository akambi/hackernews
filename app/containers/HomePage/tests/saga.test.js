/**
 * Tests for HomePage sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { LOAD_STORIES } from 'containers/App/constants';
import { storiesLoaded, repoLoadingError } from 'containers/App/actions';

import hackerNewsData, { getStories } from '../saga';

const searchQuery = 'test';

/* eslint-disable redux-saga/yield-effects */
describe('getStories Saga', () => {
  let getStoriesGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getStoriesGenerator = getStories();

    const selectDescriptor = getStoriesGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getStoriesGenerator.next(searchQuery).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the storiesLoaded action if it requests the data successfully', () => {
    const response = [
      {
        name: 'First repo',
      },
      {
        name: 'Second repo',
      },
    ];
    const putDescriptor = getStoriesGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(storiesLoaded(response, searchQuery)));
  });

  it('should call the repoLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getStoriesGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(repoLoadingError(response)));
  });
});

describe('hackerNewsDataSaga Saga', () => {
  const hackerNewsDataSaga = hackerNewsData();

  it('should start task to watch for LOAD_STORIES action', () => {
    const takeLatestDescriptor = hackerNewsDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_STORIES, getStories));
  });
});
