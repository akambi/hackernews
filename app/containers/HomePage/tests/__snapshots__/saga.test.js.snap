// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getStories Saga should call the repoLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getStories Saga should call the repoLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://hn.algolia.com/api/v1/search_by_date?tags=story",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getStories Saga should dispatch the storiesLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getStories Saga should dispatch the storiesLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://hn.algolia.com/api/v1/search_by_date?tags=story",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
